crates/red_knot_python_semantic/src/types/call.rs

```diff
@@ -1,210 +1,42 @@
 use crate::Db;
 
 mod arguments;
 mod bind;
 pub(super) use arguments::{Argument, CallArguments};
-pub(super) use bind::{bind_call, CallBinding};
+pub(super) use bind::Bindings;
 
-/// A successfully bound call where all arguments are valid.
-///
-/// It's guaranteed that the wrapped bindings have no errors.
+/// Wraps a [`Bindings`] for an unsuccessful call with information about why the call was
+/// unsuccessful.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub(super) enum CallOutcome<'db> {
-    /// The call resolves to exactly one binding.
-    Single(CallBinding<'db>),
-
-    /// The call resolves to multiple bindings.
-    Union(Box<[CallBinding<'db>]>),
-}
-
-impl<'db> CallOutcome<'db> {
-    /// Calls each union element using the provided `call` function.
-    ///
-    /// Returns `Ok` if all variants can be called without error according to the callback and `Err` otherwise.
-    pub(super) fn try_call_union<F>(
-        db: &'db dyn Db,
-        union: UnionType<'db>,
-        call: F,
-    ) -> Result<Self, CallError<'db>>
-    where
-        F: Fn(Type<'db>) -> Result<Self, CallError<'db>>,
-    {
-        let elements = union.elements(db);
-        let mut bindings = Vec::with_capacity(elements.len());
-        let mut errors = Vec::new();
-        let mut all_errors_not_callable = true;
-
-        for element in elements {
-            match call(*element) {
-                Ok(CallOutcome::Single(binding)) => bindings.push(binding),
-                Ok(CallOutcome::Union(inner_bindings)) => {
-                    bindings.extend(inner_bindings);
-                }
-                Err(error) => {
-                    all_errors_not_callable &= error.is_not_callable();
-                    errors.push(error);
-                }
-            }
-        }
-
-        if errors.is_empty() {
-            Ok(CallOutcome::Union(bindings.into()))
-        } else if bindings.is_empty() && all_errors_not_callable {
-            Err(CallError::NotCallable {
-                not_callable_type: Type::Union(union),
-            })
-        } else {
-            Err(CallError::Union(UnionCallError {
-                errors: errors.into(),
-                bindings: bindings.into(),
-                called_type: Type::Union(union),
-            }))
-        }
-    }
-
-    /// The type returned by this call.
-    pub(super) fn return_type(&self, db: &'db dyn Db) -> Type<'db> {
-        match self {
-            Self::Single(binding) => binding.return_type(),
-            Self::Union(bindings) => {
-                UnionType::from_elements(db, bindings.iter().map(CallBinding::return_type))
-            }
-        }
-    }
-
-    pub(super) fn bindings(&self) -> &[CallBinding<'db>] {
-        match self {
-            Self::Single(binding) => std::slice::from_ref(binding),
-            Self::Union(bindings) => bindings,
-        }
-    }
-}
+pub(super) struct CallError<'db>(pub(super) CallErrorKind, pub(super) Box<Bindings<'db>>);
```

Comment:
nit: I think I'd prefer named fields?

```suggestion
pub(super) struct CallError<'db> {
    pub(super) kind: CallErrorKind,
    pub(super) bindings: Box<Bindings<'db>>,
}
```

Reply:
Normally I agree but this type is only ever consumed in pattern matches, and my thought was that

```rust
Err(CallDunderError::CallError(CallErrorKind::BindingError, bindings)) => ...
```

is easier to read than

```rust
Err(CallDunderError::CallError {
    kind: CallErrorKind::BindingError,
    bindings,
})) => ...
```